### ISSUE TITLE
Better insight of cache usage.

### ISSUE THREAD

**Issue by bearbytes** [earthly/earthly#72]
Hi,

so I configured Earthly to use a cache size of 100gb, which should take me a long way, but unfortunately only lasts me for a week or so before I have to clear the cache again.

I'd like to figure out what is actually filling up my cache, but I can't seem to find any way to gain more insight.

I suggest adding commands to help people optimize their builds or decide when a cache clean is necessary:
1. `earth cache du/df` could list how much space is used/left in the build cache, making it possible to automatically clear the cache when a threshold is reached.*
2. `earth cache ls` could list the individual docker layers and mount posts command that lists the cached layers and mount points with their respective sizes, giving even more insight about where exactly the disk usage is coming from.

(* I understand that one could also use `du -s /tmp/earthly`, however this is difficult when `earth` is called from within a container that mounts the `docker.sock` of the host system. This is a scenario which often happens when e.g. Jenkins is running in a container and using Earthly to run builds. In this constellation, the cache files will be written to the host file system, not the container file system, making it inaccessible to the container).

Again, good work with this project, I do see a great future ahead for it :)

**Comment by vladaionescu** [comment]
There is possibly a way to use `buildctl` ([buildkit](https://github.com/moby/buildkit)'s CLI) to inspect the cache contents.

You might do something like

```
buildctl --addr=docker-container://earthly-buildkitd du -v
```

Though depending on circumstances, that would probably not reveal much.

----------

There are sometimes strange behaviors when buildkit is used in certain environments. If buildkit believes the underlying file system does not have copy-on-write capabilities, it ends up duplicating the base image on every single layer (rather than using copy-on-write efficiently). This causes massive disk usage. I don't know what your build entails, but $100gb definitely feels on the very high side. Maybe this could be a theory that explains it.

Given this part of your explanation:

> earth is called from within a container that mounts the docker.sock of the host system. This is a scenario which often happens when e.g. Jenkins is running in a container and using Earthly to run builds

Can you give me a few more details? Interested in the following:

* Is there any docker-in-docker at any layer in this setup? Is Jenkins itself a container, which then itself runs containers? Docker-in-docker would be the most important culprit for the theory I gave above. (For clarity: by docker-in-docker I mean running a full daemon within a container, not just mounting `docker.sock`)
* Would be helpful to understand the stack from the container that invokes earthly to the bare metal.
* The buildkit daemon that earth starts uses a mount to store the data. The source of the mount is `/tmp/earthly` on the host as you have discovered yourself. What type of filesystem is `/tmp` for your system, out of curiosity? ext4? Something else?
* What is in your `/tmp/earthly`? (Would expect to find an empty dir called `buildkit` and a file named `buildkit.img`, if this differs in any way would be good to know - eg `buildkit` dir not empty)

**Comment by vladaionescu** [comment]
Oh, also

* What [storage driver](https://docs.docker.com/storage/storagedriver/select-storage-driver/) does your docker daemon use? Maybe paste whole output of your `docker info`?

**Comment by bearbytes** [comment]
First of all, here's `docker info`:
```
Client:
 Debug Mode: false

Server:
 Containers: 18
  Running: 18
  Paused: 0
  Stopped: 0
 Images: 338
 Server Version: 19.03.6
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 
 runc version: 
 init version: 
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 4.15.0-99-generic
 Operating System: Ubuntu 18.04.4 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 7.789GiB
 Name: web-dev-server
 ID: RECF:J7AO:OGGK:WD3R:7GSK:5UZ7:YIX6:LN4Q:VTQI:5FEY:4ENI:RVZ3
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```
--- 
> Is there any docker-in-docker at any layer in this setup? Is Jenkins itself a container, which then itself runs containers? 

No and yes. Jenkins is a container, but mounts `/var/run/docker.sock` from the Host system, and uses the Host's docker daemon to run containers. So no Docker in Docker.

> Would be helpful to understand the stack from the container that invokes earthly to the bare metal.

It's quite simple:
Hardware > Ubuntu 18.04 > `jenkinsci/blueocean` container` > `earth` command
Earthly is only installed within the container, but is working with the Hosts docker daemon, as described above.

> What type of filesystem is /tmp for your system, out of curiosity? ext4? Something else?

ext4        (rw,relatime,data=ordered)

> What is in your /tmp/earthly?

Just like you described. An empty `buildkit` directory and a larger `buildkit.img` file (currently 37gb).

---

My current solution for the caching issue is to mount the `/tmp/earthly` folder from Host to the Jenkins container (read-only) and check the disk usage of that folder in a pre-build step to decide whether to prune the Build cache:

```
    stage('Check Build Cache Usage') {
      steps {
        script {
           def du = sh(script: "du -sBG /tmp/earthly", returnStdout: true).trim()
           def match = du =~ /(\d+).*/
           def gb = match[0][1].toInteger()
           echo "Build cache uses ${gb}gb of disk space"
           // Clear cache when we exceed 75gb
           if (gb > 75) {
             sh "earth prune --reset"
           }
        }
      }
    }
```

This makes it also possible to monitor how fast the cache size is growing. Currently, this is about 1 to 2 gigabytes per build, which still seems to be quite a lot. (I'm building Maven Project and an Expo App, whose `node_modules` folder is 1.4gb on my machine).

I am caching the node_modules between builds like this:
```
RUN \
  --mount=type=cache,target=./node_modules \
  npm ci --unsafe-perm
```

Which does speed up the build considerably, but might have some issues with growing cache size.

**Comment by vladaionescu** [comment]
Thank you for the info. Everything looks good - nothing to sustain my theory about possible copy-on-write issues. All that's left is probably the build just being big. I think the core issue is buildkit GC not being aggressive enough.

Can you also try using an absolute path for the cache mount target? (Note that it will end up needing to rebuild that cache because of path change.) This probably isn't an issue, but worth a shot.

----------

There is one last thing that we can do to make this work better for you. Currently, the buildkit cache is held in a loop device, which is mounted within the buildkit daemon from the file `/tmp/earthly/buildkit.img`. This loop device is there to ensure that the underlying fs is ext4 so that copy-on-write can be used. However, given that your fs is already ext4, this is not necessary so we could try to turn off this loop device and just use the underlying fs directly.

The disadvantage, however, is that the cache size would not be bound to 100GB so aggressively (ie would not be physically limited by the fs size). It is still bounded by the buildkit settings, but I've seen cache size often go over those settings when GC is not aggressive enough. Assuming that this would be ok for you, we can try this.

I'll come back with instructions later today.

**Comment by vladaionescu** [comment]
Hi @bearbytes - I've now released Earthly `v0.1.3`, which has the added config to optionally disable the loop device.

1. Download the [latest `earth`](https://github.com/earthly/earthly#linux)
2. Set env var `EARTHLY_NO_LOOP_DEVICE=true` globally (all runs of `earth` will need to have this set)

Let me know how this goes!

**Comment by vladaionescu** [comment]
Closing this issue as obsolete - the caching has improved dramatically since `v0.1.x`.
